profile/benchmarking/complex.rb

```diff
@@ -0,0 +1,104 @@
+module Elasticsearch
+  module Benchmarking
+
+    # Class encapsulating all settings and functionality for running benchmarking
+    #   tests making complex requests.
+    #
+    # @since 7.0.0
+    class Complex
+
+      # Create a Complex request benchmark test.
+      #
+      # @example Create a test.
+      #   Benchmarking::Complex.new(:patron)
+      #
+      # @param [ Symbol ] adapter The adapter the client should be configured with.
+      #
+      # @since 7.0.0
+      def initialize(adapter = ::Faraday.default_adapter)
+        @adapter = adapter
+      end
+
+      # Run a Complex request benchmark test.
+      #
+      # @example Run a test.
+      #   Benchmarking::Complex.run(:ping)
+      #
+      # @param [ Symbol ] type The name of the test to run.
+      # @param [ Integer ] repetitions The number of test repetitions.
+      #
+      # @return [ Numeric ] The test results.
+      #
+      # @since 7.0.0
+      def run(type, repetitions = TEST_REPETITIONS)
+        puts "#{type} : #{send(type, repetitions)}"
+      end
+
+      # Test sending a bulk request to create a large number of documents.
+      #
+      # @example Test sending a bulk create request.
+      #   Benchmarking::Complex.create_documents(10)
+      #
+      # @param [ Integer ] repetitions The number of test repetitions.
+      #
+      # @since 7.0.0
+      def create_documents(repetitions)
+        results = Benchmarking.with_cleanup(client) do
+          repetitions.times.collect do
+            Benchmark.realtime do
+              client.bulk(body: dataset)
+            end
+          end
+        end
+        Benchmarking.median(results)
+      end
+
+      # Test sending a request a search request.
+      #
+      # @example Test sending a search request.
+      #   Benchmarking::Complex.search_documents(10)
+      #
+      # @param [ Integer ] repetitions The number of test repetitions.
+      #
+      # @since 7.0.0
+      def search_documents(repetitions)
+        results = Benchmarking.with_cleanup(client) do
+          client.bulk(body: dataset)
```

Comment:
This benchmark is called `search_documents`, but it does insert documents are part of the run?

Reply:
No, the only measured code is in the block passed to `Benchmark.realtime` below. See here: https://github.com/elastic/elasticsearch-ruby/pull/565/files/0114fea03413df2ca61a1981ff706b3cdccc05a5#diff-35a246100c089f0ae2e3f9feb0212895R71

The dataset is only inserted once, then the search operation is run `repetitions` number of times.

Reply:
Argh, right, I conflated `Benchmarking` (the custom module) with `Benchmark` (the stdlib module).

Reply:
The naming is indeed confusing. I'll explore how to make it clearer, thanks!